Adding serialize and resume to guardianJs transactions

joseluisdiaz opened the pull request:
- new factory for `client` (websocket/pooling)
- new factory for `transaction` (fromStartFlow/fromTransactionState)
- new `.serialize` method on Transaction
- new `.resume` method in guardianJS to continue from a seralized transaction.

nikolaseu commented on README.md:
````diff
@@ -292,6 +292,35 @@ auth0GuardianJS.start(function(err, transaction) {
+```js
+var serializedTransaction = secureStorage.get('guardiantx');
+
+auth0GuardianJS.start({ transport: 'pooling' }, serializedTransaction, function(err, transaction) {
````
resume

nikolaseu commented on README.md:
````diff
@@ -292,6 +292,35 @@ auth0GuardianJS.start(function(err, transaction) {
 ```
+### auth0GuardianJS.resume(options, transactionState, callback) 
+This continues a transaction saved by `transaction.serialize()`. The option parameter provides 
+the final user the opportunity to specify which kind of `client` to use. That been:
````
`client` ?? what's client? later it's used as transport: `{ transport: 'pooling' }`

Something like this might sound better:
> The option parameter provides the final user the opportunity to specify which kind of `transport` to use, the options being:

nikolaseu commented on README.md:
````diff
@@ -389,6 +418,12 @@ recovery code is the new recovery code you must show to the user from him to sav
 ```
 
+#### transaction.serialize() 
+The .serialize() method creates a regular javascript Object that should remind opaque 
````
a "plain" object, or something better? A javascript object might have functions and other things in it. Maybe a plain JSON Object?

joseluisdiaz replied:
I like `plain`

nikolaseu commented on lib/index.js:
```diff
@@ -48,16 +44,14 @@ function auth0GuardianJS(options) {
-      self.socketClient = socketClient(self.serviceUrl);
-    }
-  }
+  self.socketClient = clientFactory({
```
`clientFactory.create` will look better, just like `transactionFactory.fromXX`
otherwise the require should have a better name, like:
```
var createClient = require('./utils/client_factory');
```
so the function call is more understandable

nikolaseu commented on lib/index.js:
```diff
@@ -128,6 +122,70 @@ auth0GuardianJS.prototype.start = function start(callback) {
+        return;
+      }
+
+      callback(null, tx);
```
```
       try {
         const tx = transactionFactory.fromTransactionState(transactionState, {
           transactionEventsReceiver: self.socketClient,
           httpClient: self.httpClient
         });
         callback(null, tx);
       } catch (transactionBuildingErr) {
         callback(transactionBuildingErr);
       }
```

joseluisdiaz replied:
?

nikolaseu commented on lib/transaction/index.js:
```diff
@@ -24,7 +24,7 @@ var eventSequencer = require('../utils/event_sequencer');
 
 /**
- * @public
+ * @public||
```
?

nikolaseu commented on lib/transaction/index.js:
```diff
@@ -74,6 +74,7 @@ function transaction(data, options) {
   self.txId = data.transactionToken.getDecoded().txid;
   self.transactionToken = data.transactionToken;
   self.transactionEventsReceiver = options.transactionEventsReceiver;
+
```
?

nikolaseu commented on lib/transaction/index.js:
```diff
@@ -136,8 +137,34 @@ function transaction(data, options) {
   return self;
 }
 
+
```
?

nikolaseu commented on lib/utils/client_factory.js:
```diff
@@ -0,0 +1,20 @@
+var pooling = require('./polling_client');
+var socketio = require('./socket_client');
+
+module.exports = function factory(options) {
```
export an object with a function `create` or something like that instead of the function directly. otherwise client_factory is not a good name

joseluisdiaz commented:
@nikolaseu addressed most of your comments; i realized that we needed a new test for `guardianjs.resume`. I rewrite the commits history, sorry about this.

fiddur commented on README.md:
````diff
@@ -389,6 +418,12 @@ recovery code is the new recovery code you must show to the user from him to sav
 ```
 
+#### transaction.serialize() 
+The .serialize() method creates a plain javascript Object that should remind opaque 
````
s/remind/remain/

fiddur replied:
...I'm not sure what is meant here, with the object being opaque to the end user.  In all examples, the object is not sent to the end user.

joseluisdiaz replied:
That's a good point; to 'library user' or 'library client' i meant.

topherMarie commented on README.md:
```diff
@@ -292,6 +292,35 @@ auth0GuardianJS.start(function(err, transaction) {
+the final user the opportunity to specify which kind of `transport` to use. That been:
+
+- `socket`: a socket.io transport
+- `pooling`: a pooling transport.
```
"The option parameter" -> "The options parameter"

"That been:" -> "Options include:"

topherMarie commented on lib/index.js:
```diff
@@ -128,6 +122,70 @@ auth0GuardianJS.prototype.start = function start(callback) {
+/**
+ * @public
+ *
  * Post result to url using an standard form
```
"Post result to url using a standard form"

topherMarie commented on lib/transaction/factory.js:
```diff
@@ -0,0 +1,76 @@
+ * @param {object} transactionState.availableAuthenticationMethods
+ *
+ * @param {EventEmitter} options.transactionEventsReceiver
+ *  Receiver for transaction events; it will receive  backend related transaction events
```
Extra space

topherMarie commented on lib/index.js:
```diff
@@ -128,6 +122,70 @@ auth0GuardianJS.prototype.start = function start(callback) {
  * @public
  *
+ * This takes a parameter what transaction.serialize returns to
+ * resume the transaction with auth0-mfa-api
```
"Takes a parameter returned from transaction.serialize to resume the transaction with auth0-mfa-api"

Correct me if I'm not interpreting it correctly, though

topherMarie commented on README.md:
```diff
@@ -292,6 +292,35 @@ auth0GuardianJS.start(function(err, transaction) {
+- `socket`: a socket.io transport
+- `pooling`: a pooling transport.
+
+The default one is `socket`. 
```
"If not set, the `socket` transport is used as default"

fiddur replied:
And `polling` again :)

fiddur commented on lib/utils/client_factory.js:
```diff
@@ -0,0 +1,20 @@
+var pooling = require('./polling_client');
```
polling!
